src/doc/trpl/the-stack-and-the-heap.md

```diff
@@ -1,3 +1,526 @@
+program. But it’s effectively unlimited in size, and is globally accessible.
+
+# The Stack
+
+Let’s talk about this Rust program:
+
+```rust
+fn main() {
+    let x = 42;
+}
+```
+
+This program has one variable binding, `x`. This memory needs to be allocated
+from somewhere. Rust ‘stack allocates’ by default, which means that basic
+values ‘go on the stack’. What does that mean?
+
+Well, when a function gets called, some memory gets allocated for all of its
+local variables and some other information. This is called a ‘stack frame’, and
+for the purpose of this tutorial, we’re going to ignore the extra information
+and just consider the local variables we’re allocating. So in this case, when
+`main()` is run, we’ll allocate a single 32-bit integer for our stack frame.
+This is automatically handled for you, as you can see, we didn’t have to write
+any special Rust code or anything.
+
+When the function is over, its stack frame gets deallocated. This happens
+automatically, we didn’t have to do anything special here.
+
+That’s all there is for this simple program. The key thing to understand here
+is that stack allocation is very, very fast. Since we know all the local
+variables we have ahead of time, we can grab the memory all at once. And since
+we’ll throw them all away at the same time as well, we can get rid of it very
+fast too.
+
+The downside is that we can’t keep values around if we need them for longer
+than a single function. We also haven’t talked about what that name, ‘stack’
+means. To do that, we need a slightly more complicated example:
+
+```rust
+fn foo() {
+    let y = 5;
+    let z = 100;
+}
+
+fn main() {
+    let x = 42;
+
+    foo();
+}
+```
+
+This program has three variables total: two in `foo()`, one in `main()`. Just
+as before, when `main()` is called, a single integer is allocated for its stack
+frame. But before we can show what happens when `foo()` is called, we need to
+visualize what’s going on with memory. Your operating system presents a view of
+memory to your program that’s pretty simple: a huge list of addresses, from 0
+to a large number, representing how much RAM your computer has. For example, if
+you have a gigabyte of RAM, your addresses go from `0` to `1,073,741,824`. That
+number comes from `2 ^ 30`, the number of bytes in a gigabyte.
+
+This memory is kind of like a giant array: addresses start at zero and go
+up to the final number. So here’s a diagram of our first stack frame:
+
+| Address | Name | Value |
++---------+------+-------+
+| 0       | x    | 42    |
+
+We’ve got `x` located at address `0`, with the value `42`.
+
+When `foo()` is called, a new stack frame is allocated:
+
+| Address | Name | Value |
++---------+------+-------+
+| 2       | z    | 100   |
+| 1       | y    | 5     |
+| 0       | x    | 42    |
+
+Because `0` was taken by the first frame, `1` and `2` are used for `foo()`’s
+stack frame. It grows upward, the more functions we call.
```

Comment:
Reading this, there is a potentially interesting ambiguity, especially for a reader who is a true novice to Rust, but not to low-level programming.

Namely: are these addresses *really* separated by one unit each (`0`, `1`, `2`, ...)?
* One obvious response is "probably not, since we are (probably) on a byte-addressed machine, which means that each address can potentially refer to an individual byte of memory, and thus each increment here represents a shift of a single byte, but our values are more than one byte apart, right?" 
 * But, ah ha, the Rust novice, forgetting how integer-fallback works, might think "oh, in the example as we have seen it so far, all of the numbers have been less than or equal to 100.  Each of those can fit into a byte, so maybe the addresses really do go up by one byte at a time."

  * (Is it strange to imagine a person reaching such a conclusion?  I'm all about anticipating the worst case for human mental models...)

Anyway, I suggest that you add a parenthetical note here that:
* The numbers `0`, `1`, and `2` are all solely for illustrative purposes, and bear no relationship to the actual numbers the computer will actually use. 
* In particular, the series of addresses are in reality going to be separated by some number of bytes that separate each address, and that separation may even exceed the size of the value being stored.

----

### Alternative
Or perhaps better still, bite the bullet:
 1. First change the original example to use explicit `i32` and `i64` type annotations, and then
 2. then use a somewhat more realistic sequence of addresses here, like `1000, 1004, 1008, ...`,
 3. (and then be sure to use an appropriate value when discussing pointer-sized types like `Box` below).

So I guess I would make the example look like:

```rust
fn foo() {
    let y: i64 = 5;
    let z: i32 = 100;
}

fn main() {
    let x: i32 = 42;

    foo();
}
```

and make the table then look like:

```
| Address | Name | #Bytes | Value |
+---------+------+--------+-------+
| 1012    | z    | 4      | 100   |
| 1004    | y    | 8      | 5     |
| 1000    | x    | 4      | 42    |
```

(even here I would still probably note that the addresses listed probably do not match what one will see in practice, since there is likely to be e.g. padding words between the two stack frames to ensure that each frame meets some expected alignment, depending on the target ABI, e.g. 16-byte alignment.)